packages/devtools_app/lib/src/charts/chart_controller.dart

```diff
@@ -0,0 +1,449 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+import 'package:intl/intl.dart' as intl;
+
+import '../auto_dispose.dart';
+
+import 'chart_trace.dart';
+
+/// Zoom level of chart
+enum displayDuration { live, fiveMinutes, fifteenMinutes, all }
+
+/// Given a durationLabel return the displayDuration.
+displayDuration durationToDisplayDuration(
+  displayDuration currentZoomDuration,
+  Duration duration,
+) {
+  if (currentZoomDuration == displayDuration.live) {
+    if (duration.inMinutes == 0 && duration.inSeconds >= 10) {
+      return displayDuration.live;
+    }
+  } else if (currentZoomDuration == displayDuration.fifteenMinutes) {
+    if (duration.inMinutes >= 5) {
+      return displayDuration.fifteenMinutes;
+    }
+  } else if (currentZoomDuration == displayDuration.fiveMinutes) {
+    if (duration.inMinutes == 3) {
+      return displayDuration.fiveMinutes;
+    }
+  }
+
+  return null;
+}
+
+String prettyTimestamp(int timestamp) {
+  final timestampDT = DateTime.fromMillisecondsSinceEpoch(timestamp);
+  return intl.DateFormat.Hms().format(timestampDT); // HH:mm:ss
+}
+
+/// Indexes into xAxisLabeledTimestamps.
+const leftLabelIndex = 0;
+const centerLabelIndex = 1;
+const rightLabelIndex = 2;
+
+/*
+                                          xPaddingRight
+                                          |
```

Comment:
what is the difference between xPaddingRight and rightPadding?

Reply:
They're both used to give space to the labels on the X-axis e.g.,
rightPadding is a constant entire Chart padding on right-side it's a constant (or could be settable) - might be better to used edget inset.  Good to add a TODO here.
xPaddingRight is a computed position of the right-side of the X-axis gives us room to display a label.